### DISCUSSION TITLE
`ReflectMetadataProvider` → Please provide either 'type' or 'entity' attribute in User.id

### DISCUSSION THREAD

**Discussion by maximedupre** [mikro-orm/mikro-orm#4278] [Q&A] [answered]
**Describe the bug**
It seems like the annotations in my `users.ts` file are not working...

``` typescript
import { Entity, PrimaryKey, Property } from '@mikro-orm/core';

@Entity()
export default class User {
    @PrimaryKey()
    id?: number;

    @Property()
    email: string;

    @Property()
    password: string;

    constructor(email: string, password: string) {
        this.email = email;
        this.password = password;
    }
}
```

My `tsconfig.json` file has the necessary configs:

``` json
...
"experimentalDecorators": true,
"emitDecoratorMetadata": true,
"esModuleInterop": true
```

**Stacktrace**
```
Please provide either 'type' or 'entity' attribute in User.id
Error: Please provide either 'type' or 'entity' attribute in User.id
    at ReflectMetadataProvider.initPropertyType (./node_modules/@mikro-orm/core/metadata/ReflectMetadataProvider.js:14:19)
    at ./node_modules/@mikro-orm/core/metadata/ReflectMetadataProvider.js:9:54
    at ReflectMetadataProvider.initProperties (./node_modules/@mikro-orm/core/metadata/MetadataProvider.js:38:23)
    at ReflectMetadataProvider.loadEntityMetadata (./node_modules/@mikro-orm/core/metadata/ReflectMetadataProvider.js:9:20)
    at MetadataDiscovery.discoverEntity (./node_modules/@mikro-orm/core/metadata/MetadataDiscovery.js:236:37)
    at MetadataDiscovery.discoverReferences (./node_modules/@mikro-orm/core/metadata/MetadataDiscovery.js:175:24)
    at MetadataDiscovery.findEntities (./node_modules/@mikro-orm/core/metadata/MetadataDiscovery.js:91:20)
    at async MetadataDiscovery.discover (./node_modules/@mikro-orm/core/metadata/MetadataDiscovery.js:34:9)
    at async MikroORM.discoverEntities (./node_modules/@mikro-orm/core/MikroORM.js:115:25)
    at async MySqlMikroORM.init (./node_modules/@mikro-orm/core/MikroORM.js:34:9
```

**Additional context**
I'm using `vite` for my builds.

There is this issue: https://github.com/mikro-orm/mikro-orm/issues/841, but then the "solution" in this ticket is to use `TsMorphMetadataProvider` instead, which is not a solution per se if you want to use `ReflectMetadataProvider`.

**Versions**

| Dependency | Version |
| - | - |
| node | 18.13.0 |
| typescript | ^5.0.0 |
| mikro-orm | ^5.7.1 |
| @mikro-orm/mysql | ^5.7.1 |

**Comment by B4nan** [comment]
> I'm using vite for my builds.

Most probably caused by that, not sure if they support decorator metadata somehow, quick search pointed me [here](https://stackoverflow.com/questions/68570519/why-cant-reflect-metadata-be-used-in-vite), where it says they use esbuild to compile TS - and esbuild has no support for decorator metadata.

**Reply by maximedupre** [reply]
Hey Martin, thanks for the quick answer. So if I'm using Vite, which uses esbuild under the hood, which does not support decorator metadata, can you help me list my options if I want to keep using MikroORM? Those are the options I see right now:

1. Keep using `ReflectMetadataProvider` and specify the `type` attribute for all `@Property()` annotations
2. Use `TsMorphMetadataProvider` instead (do you see a potential conflict between esbuild and `TsMorphMetadataProvider`?)
3. Use `JavaScriptMetadataProvider` instead (can I still write my code as TypeScript even when using this metadata provider?)

Thanks 🙏

**Reply by B4nan** [reply, marked as answer]
I would go with ts-morph, the only reason it was considered problematic for some was the deployment and the need to have TS as runtime dependency - which is already resolved in v6, where you can depend on it only as a dev dependency, create metadata cache during build and use that in your production build (see #4164 and #4167).

There are quite a lot of hidden improvements in v6 that are targetting new school tools like vite, e.g. #3814. If you are starting a new project or just trying things out, try v6 (upgraing guide is [here](https://github.com/mikro-orm/mikro-orm/blob/v6/docs/docs/upgrading-v5-to-v6.md)).

> Those are the options I see right now:

Yes, that sounds about right, just the 3. point should be using [`EntitySchema`](https://mikro-orm.io/docs/entity-schema), the `JavaScriptMetadataProvider` is deprecated and already removed from v6. I don't thing there are any issues with esbuild and ts-morph, FYI the new getting started guide I am working on (ok, more like postponing to work on it) is using vitest (so esbuild too) with ts-morph.

**Reply by B4nan** [reply]
Plus the 4. option would be the one mentioned in that SO answer I originally linked, it says it should be possible to use swc with vite too, and swc has support for decorator metadata.

**Reply by B4nan** [reply]
I've invited you to the getting started guide repo, maybe you will find that useful. It's far away from ready (hence its still private) but the first half is pretty much done and explains the concepts in a better way than current docs (I hope).

Feedback on that welcome!

**Reply by maximedupre** [reply]
The guide is really useful so far, especially since it uses ES modules and `TsMorphMetadataProvider`. 

I'm having issues with the discoverability of my entities: 

```
Source file '../../../../../../src/lib/server/entities/user.ts' not found. Check your 'entitiesTs' option and verify you have 'compilerOptions.declaration' enabled in your 'tsconfig.json'. If you are using webpack, see https://bit.ly/35pPDNn
```

So I guess I might have to do some convoluted stuff in my Vite build chain 😁

But you said that v6 is better fitted for "new school tools like vite", so I thought it would be better to upgrade before going into the rabbit hole of trying to figure out what's wrong. 

So how does one upgrade to v6? Your [upgrade guide](https://github.com/mikro-orm/mikro-orm/blob/v6/docs/docs/upgrading-v5-to-v6.md) seems to explain the breaking changes in depth, but not how to actually upgrade. Maybe it should be obvious how to upgrade 😅, but what came to mind for me was to instead using @ next, but that points to `5.7.3-dev.5`. How should I think about installing the package? Should I install the npm package pointing to the v6 Github branch? 🤔 (there doesn't seem to be any v6 release/tag)

Thanks 🙏

**Edit**

Btw I'm specifying my entities directly like this: `entities: [User]`. It thought it would avoid path issues and stuff, but I guess not 🥹

**Reply by B4nan** [reply]
There is `next-v6` tag on NPM:

<img width="787" alt="image" src="https://user-images.githubusercontent.com/615580/234908970-5ad9d66e-69ad-434d-8414-8af671fb0008.png">

I will switch it to `next` once RC version is released probably.

> Btw I'm specifying my entities directly like this: entities: [User]. It thought it would avoid path issues and stuff, but I guess not 🥹

ts-morph will always work based on the source files, it needs to process the TS source code to get the metadata - you don't have them on runtime anywhere. Maybe you need to set the `baseDir`? It defaults to the current working directory.

https://mikro-orm.io/docs/configuration#entity-discovery

**Reply by maximedupre** [reply]
Got it! I should have looked at the NPM tags as well.

I did upgrade all packages to v6:

``` json
"@mikro-orm/core": "^6.0.0-dev.64",
"@mikro-orm/mysql": "^6.0.0-dev.64",
"@mikro-orm/reflection": "^6.0.0-dev.64",
```

And provided the paths to the ts entities:

``` typescript
import { MikroORM, type MySqlDriver } from '@mikro-orm/mysql';
import { TsMorphMetadataProvider } from '@mikro-orm/reflection';
import User from './entities/user';

export const orm = await MikroORM.init<MySqlDriver>({
    entities: [User],
    entitiesTs: ['./src/lib/server/entities'],
    dbName: 'myDb',
    metadataProvider: TsMorphMetadataProvider,
    baseDir: process.cwd(), // I console logged this and it points to the right place (the root of my project)
});
```

Yet there is still the same error. I suppose I would need to pass some sort of path for the `entities` config as well, but I wouldn't know what 🤔 I'm running `vite dev`, so all the files are internal to the web dev server. I ran `vite build` to explore the output that would be used for production, and found the `User` class in this file `.svelte-kit/output/server/entries/endpoints/api/users/_server.ts.js`, obviously mixed in with some other stuff lol Do you have some pointers for what I should do next?

**Reply by B4nan** [reply]
I think that for production build you will need that `GeneratedCacheAdapter` added in #4167 (or alternatively you'd have to compile the `.d.ts` files via `tsc` next to your JS bundle). 

Try adding `tsNode: true` to the config, normally this is detected automatically when you use ts-node or ts-jest, but since you are using vite, you will need to set that manually (will need to take a closer look if this can be somehow detected too). Which reminds me I should rename this option in v6, as it's no longer only about ts-node, its more about switching to development mode.

What the `tsNode` option does is switching between `entities` and `entitiesTs` - so when its enabled, the `entitiesTs` will be used instead.

>    entities: [User],
>    entitiesTs: ['./src/lib/server/entities'],

This is kinda wrong, you should keep only the `entities`, or have both `entities` and `entitiesTs` aligned - defined in the same way. The path to the entity will be automatically detected when you use some decorator ([extracted from a stack trace](https://github.com/mikro-orm/mikro-orm/blob/master/packages/core/src/utils/Utils.ts#L694)), I think that should work with vite too, since it is working with vitest in the getting started guide. `entitiesTs` will have no effect unless the TS mode is enabled (the `tsNode` option).

> baseDir: process.cwd(),

And this is the default, so specifying it manually won't change anything :]

If you want to stay away from the file system issues, using `EntitySchema` would do it, or the default reflect-metadata provider, but you would have to specify the types yourself. Those two solutions won't touch the file system, they will force you to provide the metadata if reflection fails (exactly what you saw initially).

If you can provide some repro, I can take a look.

**Reply by maximedupre** [reply]
`tsNode: true` seems to move the ball a bit 😁 but now:

```
TypeError [ERR_UNKNOWN_FILE_EXTENSION]: Unknown file extension ".ts" for /src/lib/server/entities/user.ts
```

It would be awesome if you could take a quick look at my repro (I have invited you to my repo) 🙏
1. `npm i`
2. `npm run dev`
3. Navigate to `/register'
4. Click on "Register" → error appears on UI & in console

**Edit 1**: I tried all possible variations of `entities` & `entitiesTs` with `tsNode`
**Edit 2**: Seems like an esm issue... Normally I resolve this kind of issue in `ts-node` by passing the `--esm` flag 🤔

**Reply by B4nan** [reply]
<img width="1384" alt="image" src="https://user-images.githubusercontent.com/615580/235196629-2762c292-06ae-4aa4-b58a-e8ad0cff9316.png">

https://github.com/mikro-orm/guide#setting-up-cli

**Reply by maximedupre** [reply]
I'm not sure I understand how installing the CLI (specifically its ESM variation) is going to help me fix this issue.

As I understand, the CLI is used to make migrations, schema manipulations, etc and I'm not currently doing any of that.

I'm still trying to run my dev server using the `TsMorphMetadataProvider`. Am I missing something?
